### PR TITLE
Add option :return_errors to ParallelCompiler and ParallelRequire

### PR DESCRIPTION
**Pull request by JakeBecker**
This is motivated by the goal of allowing Mix compilers to return their diagnostic information in a standardized format. Before this change, `ParallelCompiler` calls `exit({:shutdown, 1})` upon encountering a compilation error, and there's no way to programatically retrieve the errors. This change allows `ParallelCompiler` and `ParallelRequire` to return warnings and errors if the option `:return_errors` is passed. Return values is `{:ok, modules, warnings} | {:error, errors}` when this option is passed. If the option is not passed, the return value is the same as before, and it will exit the process when encountering an error.

There seemed to be a lot of duplication between the code in `ParallelRequire` and `ParallelCompiler`, so to avoid having to duplicate these changes in both places, I added an option `:require` to `ParallelCompiler` to load files via `Code.require_files/1`. Now, `ParallelRequire` simply calls `ParallelCompiler` with this option, reducing code duplication.

### CODE REVIEW

File: lib/elixir/lib/kernel/parallel_compiler.ex [hunk 1]
```diff
@@ -56,6 +65,15 @@ defmodule Kernel.ParallelCompiler do
     spawn_compilers(files, path, options)
   end
 
+  @doc """
+  Requires the given files in parallel.
+  Read `files/2` for more information.
+  """
+  def require_files(files, options \\ []) do
```
Comment (josevalim) [inline lib/elixir/lib/kernel/parallel_compiler.ex]: The parallel compiler and parallel require have very distinct semantics. The compiler is capable of solving dependency conflicts, the parallel require is not. Maybe we should be able to unify both but that is certainly out of scope of this PR.

Why are you unifying both?

Comment (JakeBecker) [inline lib/elixir/lib/kernel/parallel_compiler.ex]: I wanted to add warning and error returns to `ParallelRequire` in order to return diagnostics from the test compiler for test files. I started adding it to `ParallelRequire` but it seemed a fair amount of code duplication. I handled the different semantics for `require` by not setting the custom error handler in `ParallelCompile` if the `require` option is set: https://github.com/elixir-lang/elixir/pull/6446/files#diff-59dfea30c24e10aaa6d3e55047e926d6R144

If you like, I'll remove the changes to `ParallelRequire` from this PR, but adding warning and error returns to `ParallelRequire` will further increase the code duplication if we don't unify them at some point.

Comment (josevalim) [inline lib/elixir/lib/kernel/parallel_compiler.ex]: Let's just change the ParallelCompiler then. If the issue is with testing, then we may even figure out a way to make `mix test` use the ParallelCompiler instead.

**Comment by JakeBecker** [comment]
@josevalim I undid the unification of `ParallelCompiler` and `ParallelRequire`. Let me know if you want me to squash to keep commit history cleaner.

**Comment by JakeBecker** [comment]
@josevalim Bump. Can I get a review on this? Adding diagnostics would let me greatly improve ElixirLS. Atom IDE was just announced, and it uses the language server protocol: https://blog.atom.io/2017/09/12/announcing-atom-ide.html , so that's another reason to make these changes to improve ElixirLS.

**Comment by josevalim** [comment]
Thank you @JakeBecker! Wouldn't it be better for the return type to be `{:ok, module, warnings} and `{:error, errors, warnings}`? This way you always get warnings, even if there are errors?

**Comment by JakeBecker** [comment]
@josevalim Sounds good to me. I made the change.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by josevalim** [comment]
Thank you @JakeBecker! I really like this change.

I have pushed another commit 4a71760e2fb8850a3df24953a92ac21d4f178376 which makes returning :ok/:error our official API and we will deprecate the older one in the long term.

Are you planning to do the same changes to ParallelRequire?

**Comment by josevalim** [comment]
@JakeBecker actually, let me see if we can unify ParallelRequire and ParallelCompiler as you proposed but keeping the require semantics.

**Comment by josevalim** [comment]
@JakeBecker unifying them worked great, thanks again!

**Pull request merged**